app/src/main/java/io/homeassistant/companion/android/widgets/WidgetDynamicFieldAdapter.kt

```diff
@@ -0,0 +1,100 @@
+package io.homeassistant.companion.android.widgets
+
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.AutoCompleteTextView
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import io.homeassistant.companion.android.domain.integration.Entity
+import io.homeassistant.companion.android.domain.integration.Service
+import kotlinx.android.synthetic.main.widget_button_configure_dynamic_field.view.*
+
+class WidgetDynamicFieldAdapter(
+    private val services: HashMap<String, Service>,
+    private val entities: HashMap<String, Entity<Any>>,
+    private val serviceFieldList: ArrayList<Pair<String, String>>
+) : RecyclerView.Adapter<WidgetDynamicFieldAdapter.ViewHolder>() {
+    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
+
+    private val fieldTextViewList = ArrayList<TextView>()
```

Comment:
This should be an ArrayList of Strings.

Reply:
So, I ended up entirely reworking this so that the text that gets saved is directly associated with the appropriate field key.  It's a bit more of a direct method I'd like you to take a look at.